datafusion/core/src/datasource/object_store.rs

```diff
@@ -132,19 +149,51 @@ impl ObjectStoreRegistry {
     ///
     /// - URL with scheme `file:///` or no schema will return the default LocalFS store
     /// - URL with scheme `s3://bucket/` will return the S3 store if it's registered
+    /// - URL with scheme `hdfs://hostname:port` will return the hdfs store if it's registered
     ///
     pub fn get_by_url(&self, url: impl AsRef<Url>) -> Result<Arc<dyn ObjectStore>> {
         let url = url.as_ref();
-        let s = &url[url::Position::BeforeScheme..url::Position::AfterHost];
-        let stores = self.object_stores.read();
-        let store = stores.get(s).ok_or_else(|| {
-            DataFusionError::Internal(format!(
-                "No suitable object store found for {}",
-                url
-            ))
-        })?;
-
-        Ok(store.clone())
+        // First check whether can get object store from registry
+        let store = {
+            let stores = self.object_stores.read();
+            let s = &url[url::Position::BeforeScheme..url::Position::BeforeHost];
```

Comment:
What if we called the detector *first* and if that does not return an instance, we can fall back to the `self.object_stores`? This would preserve the existing behavior and would likely have a simpler implementation

Something like (untested): 

```rust
let store = self_detector
  // try to get url from detector first
  .map(|detector| detector.get_by_url(url))
  // fallback to looking up by registered scheme
  .or_else(|| self.oject_stores.read().get(s))
  .ok_or_else(|| {
                DataFusionError::Internal(format!(
                    "No suitable object store found for {}",
                    url
                ))
            })?;
```

Reply:
I think it's better to call the detector later so that we can reuse the registered object stores.

Reply:
What I don't like about the current code is that it is hard to concisely explain how urls are resolved as various parts of the URL are tried in a sequence that I find confusing. 

I guess I was thinking that we could keep the logic in DataFusion simple,  Ballista or some other system could implement whatever arbitrary logic was needed in its implementation of the detector. If the detector wanted to first try the registered object stores it could do so.

Reply:
The previous logic is not correct for the hdfs case, like hdfs://localhost:8020/xxx/xxx. The key needs to include the port info.

Reply:
Oh sorry, that's my mistake - it should be BeforePath not AfterHost, agreed 👍

Reply:
If use before path, then it may include credential infos to the key for S3, right? 😭 It's also not good, right?

Reply:
Why is that not good? I'd expect to be able to register two object stores with the same host but different credentials? I don't follow why this is an issue?

Reply:
It will easily cause security issues. In general, we should not store credentials in memory, especially directly depending on it.

Reply:
Could you perhaps expand on the threat model here, and how this would impact it. If the user has opted to use HTTP Basic Auth, I don't see how storing that data in memory as part of the URL is avoidable...

Reply:
In general, we should not store the users' credential info in memory. The security topic related to many points:
- The expiration mechanism
- The encoding algorithm
- The encryption algorithm
- ...

However, this PR is not for dealing with it. If anyone is interested in it, we should open another issue. For HDFS, we leverage Kerberos technique to deal with security issue. However, I have little knowledge for S3. My intuition tells me that it's not safe to put the credential info in each request, although it's base64 encoded.

For this PR, it's OK to just use *url::Position::BeforeScheme..url::Position::BeforePath* to extract the key for one specific object store. And I'll submit a related commit to fix it.